### PR TITLE
avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`.

### PR DESCRIPTION
**Pull request by gouhongshen**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #13013 #13151 #12673 #12994

## What this PR does / why we need it:
1. adding debug log to dump catalog cache  when `ExpectedEOB` for #13151 
2. avoid loading object meta when object stats are valid when `mo_table_size` and `metadata_scan`.
> For `metadata_scan`, if no col is specified, col meta info will be `NULL`.

cherry-pick to 1.1: #13949

### CODE REVIEW

**Comment by matrix-meow** [comment]
@gouhongshen Thanks for your contributions!

**Here are review comments for file pkg/sql/plan/function/func_mo.go:**
Title: Avoiding load object meta when object stats are valid.

Body: This pull request addresses two issues (#13013 and #13151). It adds a debug log to dump catalog cache when `ExpectedEOB` occurs for issue #13151. Additionally, it avoids loading object meta when object stats are valid for both `mo_table_size` and `metadata_scan` functions.

Changes for file pkg/sql/plan/function/func_mo.go:
- Added a check for `ExpectedEOB` error in the `MoTableRows` function. If the error is encountered, a debug log is printed and an `InvalidArgNoCtx` error is returned.
- Added a debug log for issue #13151 in the `MoTableRows` function.
- Added a check for `ExpectedEOB` error in the `MoTableSize` function. If the error is encountered, a debug log is printed and an `InvalidArgNoCtx` error is returned.

Review:

1. In the `MoTableRows` function, the check for `ExpectedEOB` error is handled correctly. However, the debug log should be removed before merging the code to the main branch. It is not necessary to keep debug logs in the production code.

Suggestion: Remove the debug log before merging.

2. In the `MoTableSize` function, the check for `ExpectedEOB` error is handled correctly. However, the debug log should also be removed before merging.

Suggestion: Remove the debug log before merging.

Overall, the changes made in this pull request are necessary to address the mentioned issues. However, the debug logs should be removed before merging the code to the main branch to avoid unnecessary logging in the production code.

**Here are review comments for file pkg/vm/engine/disttae/cache/catalog.go:**
Title: avoiding load object meta when object stats are valid.

Body: This pull request addresses two issues (#13013 and #13151). It adds a debug log to dump catalog cache for issue #13151 and avoids loading object metadata when object stats are valid for both `mo_table_size` and `metadata_scan`.

Changes made in `pkg/vm/engine/disttae/cache/catalog.go`:
- Added a new function `TraverseDbAndTbl` to traverse the database and table cache for debugging purposes.
- Added new types `DebugTableItem` and `DebugDatabaseItem` to represent table and database items for debugging.
- Added `String` methods to `DebugTableItem` and `DebugDatabaseItem` for string representation.
- Added `TraverseTableCache` and `TraverseDatabaseCache` methods to traverse the table and database cache respectively and return debug items.

Overall, the changes seem to be focused on debugging and improving the catalog cache functionality.

Potential problems and suggestions for improvement:
1. The added debug log in `TraverseDbAndTbl` should be removed before merging the code. It is mentioned in the comments that it will be removed later, but it should be removed before merging to avoid unnecessary debug logs in production code.
2. The new types `DebugTableItem` and `DebugDatabaseItem` could benefit from additional documentation to explain their purpose and usage.
3. The `String` methods in `DebugTableItem` and `DebugDatabaseItem` could be improved by using a more descriptive format for the string representation. For example, including labels for each field in the format string.
4. It is not clear why the `TraverseTableCache` and `TraverseDatabaseCache` methods are added as separate methods instead of being part of the existing `tableCache` and `databaseCache` structs. It might be worth considering refactoring the code to make these methods part of the respective structs for better organization and encapsulation.
5. It would be helpful to have some unit tests for the new functions and methods added in this pull request to ensure their correctness and prevent regressions in the future.

Optimization suggestions:
1. The `TraverseTableCache` and `TraverseDatabaseCache` methods could potentially be optimized by pre-allocating the `ret` slice with the correct capacity based on the number of items in the cache. This would avoid unnecessary reallocations and improve performance, especially if the cache is large.
2. Consider using a more efficient data structure for the cache, such as a map, to improve lookup and traversal performance. This could be beneficial if the cache is frequently accessed or updated.

Security issues:
No security issues were identified in this pull request.

**Here are review comments for file pkg/vm/engine/disttae/engine.go:**
Review:

Title: avoiding load object meta when object stats are valid.

Body: The pull request fixes two issues (#13013 and #13151) and makes two changes in the `engine.go` file. The changes include adding debug logs to dump catalog cache and avoiding loading object metadata when object stats are valid.

Changes in `engine.go`:
1. Added import statements for `bytes`, `fmt`, and `sort`.
2. Added a new function `enableDebug()` which sets a debug function for `DebugGetDatabaseExpectedEOB`.
3. Added a new function `DebugGetDatabaseExpectedEOB()` which traverses the database and table structures, sorts them, and logs the expected end of batch (EOB) information.
4. Modified the `New()` function to call `enableDebug()` for debugging purposes.
5. Modified the `Create()` function to use the `getTransaction()` method to retrieve the transaction.

Suggestions:
1. The changes made in the pull request seem to be related to debugging and logging. It would be helpful to provide more context on why these changes are necessary and how they address the identified issues.
2. The added debug functions and logging can be useful for troubleshooting, but it's important to ensure that they are properly controlled and not exposed in production environments. Consider adding a flag or configuration option to enable/disable the debug functionality.
3. The `enableDebug()` function can be simplified by directly assigning the debug function to `function.DebugGetDatabaseExpectedEOB` without the need for an intermediate method call.
4. The `DebugGetDatabaseExpectedEOB()` function can be optimized by using a `strings.Builder` instead of a `bytes.Buffer` for better performance.
5. Consider adding comments to explain the purpose and functionality of the added code.
6. It would be helpful to include test cases for the changes made in the pull request to ensure that they work as expected and to prevent regressions in the future.

**Here are review comments for file pkg/vm/engine/disttae/logtailreplay/partition_state.go:**
Review:

Title: avoiding load object meta when object stats are valid.

Body: 
- The PR fixes issues #13013 and #13151.
- The changes in this PR include adding a debug log to dump catalog cache when `ExpectedEOB` for issue #13151 and avoiding loading object meta when object stats are valid for `mo_table_size` and `metadata_scan`.

Changes in `pkg/vm/engine/disttae/logtailreplay/partition_state.go`:
- Added a new method `StatsValid()` to the `ObjectInfo` struct, which returns a boolean indicating whether the object stats are valid or not.

Overall, the changes in this PR seem to address the mentioned issues and improve the code by avoiding unnecessary loading of object meta when the object stats are already valid.

However, there are a few suggestions for improvement:

1. It would be helpful to provide more context in the PR body about why avoiding loading object meta when object stats are valid is beneficial. This could include any performance improvements or reduction in resource usage.

2. The debug log added for issue #13151 should be reviewed to ensure it provides useful information for debugging purposes. It would be helpful to include a brief explanation of what information is being logged and why it is relevant to the issue.

3. It would be beneficial to add some unit tests to cover the new `StatsValid()` method to ensure its correctness and to provide documentation for future maintainers.

4. Consider providing more information in the PR title about the specific issue being addressed, such as "Fixing issue #13151 by avoiding loading object meta when stats are valid".

Overall, the changes in this PR are positive and address the mentioned issues. The suggestions provided above will help improve the code quality and maintainability.

**Here are review comments for file pkg/vm/engine/disttae/txn_table.go:**
Review:

Title: avoiding load object meta when object stats are valid.

Body: This pull request aims to avoid loading object metadata when the object stats are valid. It also adds a debug log to dump catalog cache for issue #13151. The changes are made in the file `pkg/vm/engine/disttae/txn_table.go`.

Changes Made:
- In the `Size` function, an additional check is added to skip loading object metadata if the object stats are valid. This is done by checking if `obj.StatsValid()` returns true. If the stats are valid, the size of the object is added to the `ret` variable and the function returns.
- In the `GetColumMetadataScanInfo` function, a similar check is added to skip loading object metadata if the object stats are valid. Additionally, if the `name` parameter is set to "AllColumns", the function directly creates `MetadataScanInfo` objects using the object's stats without loading the object metadata.

Feedback:
1. In the `Size` function, it would be helpful to add a comment explaining why loading object metadata is skipped when the object stats are valid. This will make the code more understandable for future maintainers.
2. In the `GetColumMetadataScanInfo` function, it would be beneficial to add a comment explaining the purpose of the check for "AllColumns" and why loading object metadata is skipped in this case.
3. It is unclear why a debug log is added to dump catalog cache for issue #13151. The purpose and impact of this change should be clarified in the pull request description or in a comment in the code.
4. It is important to ensure that skipping the loading of object metadata when the object stats are valid does not introduce any security vulnerabilities. It would be advisable to conduct a thorough security review to ensure that this change does not compromise the integrity or confidentiality of the data.

Optimization Suggestions:
1. In the `Size` function, the check for `obj.StatsValid()` can be moved to the beginning of the function to avoid unnecessary variable assignments and function calls.
2. In the `GetColumMetadataScanInfo` function, the check for "AllColumns" can be moved to the beginning of the function to avoid unnecessary variable assignments and function calls.

Overall, the changes made in this pull request seem reasonable, but further clarification and security review are needed. Additionally, optimizing the code by moving the checks to the beginning of the functions can improve performance.

**Review by aressu1985 — APPROVED** [review]

**Review by reusee — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Pull request merged**
